File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-asset-5-4/MutualFund.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class MutualFund : UnderlyingAsset
    {
        public MutualFund(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode openEndedFundNode = xmlNode.SelectSingleNode("openEndedFund");
            
            if (openEndedFundNode != null)
            {
                if (openEndedFundNode.Attributes["href"] != null || openEndedFundNode.Attributes["id"] != null) 
                {
                    if (openEndedFundNode.Attributes["id"] != null) 
                    {
                        openEndedFundIDRef_ = openEndedFundNode.Attributes["id"].Value;
                        XsdTypeBoolean ob = new XsdTypeBoolean(openEndedFundNode);
                        IDManager.SetID(openEndedFundIDRef_, ob);
                    }
                    else if (openEndedFundNode.Attributes["href"] != null)
                    {
                        openEndedFundIDRef_ = openEndedFundNode.Attributes["href"].Value;
                    }
                    else
                    {
                        openEndedFund_ = new XsdTypeBoolean(openEndedFundNode);
                    }
                }
                else
                {
                    openEndedFund_ = new XsdTypeBoolean(openEndedFundNode);
                }
            }
            
        
            XmlNode fundManagerNode = xmlNode.SelectSingleNode("fundManager");
            
            if (fundManagerNode != null)
            {
                if (fundManagerNode.Attributes["href"] != null || fundManagerNode.Attributes["id"] != null) 
                {
                    if (fundManagerNode.Attributes["id"] != null) 
                    {
                        fundManagerIDRef_ = fundManagerNode.Attributes["id"].Value;
                        XsdTypeString ob = new XsdTypeString(fundManagerNode);
                        IDManager.SetID(fundManagerIDRef_, ob);
                    }
                    else if (fundManagerNode.Attributes["href"] != null)
                    {
                        fundManagerIDRef_ = fundManagerNode.Attributes["href"].Value;
                    }
                    else
                    {
                        fundManager_ = new XsdTypeString(fundManagerNode);
                    }
                }
                else
                {
                    fundManager_ = new XsdTypeString(fundManagerNode);
                }
            }
            
        
        }
        
    
        #region OpenEndedFund_
        private XsdTypeBoolean openEndedFund_;
        public XsdTypeBoolean OpenEndedFund_
        {
            get
            {
                if (this.openEndedFund_ != null)
                {
                    return this.openEndedFund_; 
                }
                else if (this.openEndedFundIDRef_ != null)
                {
                    openEndedFund_ = IDManager.getID(openEndedFundIDRef_) as XsdTypeBoolean;
                    return this.openEndedFund_; 
                }
                else
                {
                      return this.openEndedFund_; 
                }
            }
            set
            {
                if (this.openEndedFund_ != value)
                {
                    this.openEndedFund_ = value;
                }
            }
        }
        #endregion
        
        public string openEndedFundIDRef_ { get; set; }
        #region FundManager_
        private XsdTypeString fundManager_;
        public XsdTypeString FundManager_
        {
            get
            {
                if (this.fundManager_ != null)
                {
                    return this.fundManager_; 
                }
                else if (this.fundManagerIDRef_ != null)
                {
                    fundManager_ = IDManager.getID(fundManagerIDRef_) as XsdTypeString;
                    return this.fundManager_; 
                }
                else
                {
                      return this.fundManager_; 
                }
            }
            set
            {
                if (this.fundManager_ != value)
                {
                    this.fundManager_ = value;
                }
            }
        }
        #endregion
        
        public string fundManagerIDRef_ { get; set; }
        
    
        
    
    }
    
}

File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-asset-5-4/MutualFund.cpp
// MutualFund.cpp 
#include "MutualFund.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

MutualFund::MutualFund(TiXmlNode* xmlNode)
: UnderlyingAsset(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //openEndedFundNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* openEndedFundNode = xmlNode->FirstChildElement("openEndedFund");

   if(openEndedFundNode){openEndedFundIsNull_ = false;}
   else{openEndedFundIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- openEndedFundNode , address : " << openEndedFundNode << std::endl;
   #endif
   if(openEndedFundNode)
   {
      if(openEndedFundNode->Attribute("href") || openEndedFundNode->Attribute("id"))
      {
          if(openEndedFundNode->Attribute("id"))
          {
             openEndedFundIDRef_ = openEndedFundNode->Attribute("id");
             openEndedFund_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(openEndedFundNode));
             openEndedFund_->setID(openEndedFundIDRef_);
             IDManager::instance().setID(openEndedFundIDRef_,openEndedFund_);
          }
          else if(openEndedFundNode->Attribute("href")) { openEndedFundIDRef_ = openEndedFundNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { openEndedFund_ = boost::shared_ptr<XsdTypeBoolean>(new XsdTypeBoolean(openEndedFundNode));}
   }

   //fundManagerNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* fundManagerNode = xmlNode->FirstChildElement("fundManager");

   if(fundManagerNode){fundManagerIsNull_ = false;}
   else{fundManagerIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- fundManagerNode , address : " << fundManagerNode << std::endl;
   #endif
   if(fundManagerNode)
   {
      if(fundManagerNode->Attribute("href") || fundManagerNode->Attribute("id"))
      {
          if(fundManagerNode->Attribute("id"))
          {
             fundManagerIDRef_ = fundManagerNode->Attribute("id");
             fundManager_ = boost::shared_ptr<XsdTypeString>(new XsdTypeString(fundManagerNode));
             fundManager_->setID(fundManagerIDRef_);
             IDManager::instance().setID(fundManagerIDRef_,fundManager_);
          }
          else if(fundManagerNode->Attribute("href")) { fundManagerIDRef_ = fundManagerNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { fundManager_ = boost::shared_ptr<XsdTypeString>(new XsdTypeString(fundManagerNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<XsdTypeBoolean> MutualFund::getOpenEndedFund()
{
   if(!this->openEndedFundIsNull_){
        if(openEndedFundIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeBoolean>(IDManager::instance().getID(openEndedFundIDRef_));
        }else{
             return this->openEndedFund_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeBoolean>();
   }
}
boost::shared_ptr<XsdTypeString> MutualFund::getFundManager()
{
   if(!this->fundManagerIsNull_){
        if(fundManagerIDRef_ != ""){
             return boost::shared_static_cast<XsdTypeString>(IDManager::instance().getID(fundManagerIDRef_));
        }else{
             return this->fundManager_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<XsdTypeString>();
   }
}
}

